### PR TITLE
Issue #7853; 3D zoom to map via share URL - fix after test

### PR DESCRIPTION
**Pull request by tonym3nsa**
## Description

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

## Issue

**What is the current behavior?**

#7853 

**What is the new behavior?**

- The 3D map zooms to the location as provided in the share URL
- The share form accepts negative values for roll and pitch

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

File: web/client/components/share/SharePanel.jsx [hunk 1]
```diff
@@ -151,6 +151,13 @@ class SharePanel extends React.Component {
             !isEqual(this.props.isVisible, newProps.isVisible)) {
             this.initializeDefaults(newProps);
         }
+        if (this.props?.viewerOptions?.orientation) {
```
Comment (dsuren1) [inline web/client/components/share/SharePanel.jsx:154]: I think we can move this block inside `initializeDefaults` by adding a condition to check `!isEqual(this.props?.viewerOptions?.orientation, newProps.viewerOptions?.orientation)` invoking `initialzeDefaults` func

**Review by dsuren1 — CHANGES_REQUESTED** [review]
@tonym3nsa 
Looks fine. Kindly update the PR with this minor change.

File: web/client/components/share/SharePanel.jsx (hunk 1 above)
Comment (tonym3nsa) [reply web/client/components/share/SharePanel.jsx:154]: well noted. Thanks

File: web/client/components/share/SharePanel.jsx [hunk 2]
```diff
@@ -151,6 +151,9 @@ class SharePanel extends React.Component {
             !isEqual(this.props.isVisible, newProps.isVisible)) {
             this.initializeDefaults(newProps);
         }
+        if (!isEqual(this.props?.viewerOptions?.orientation, newProps.viewerOptions?.orientation)) {
+            this.initializeDefaults(newProps);
+        }
```
Comment (dsuren1) [inline web/client/components/share/SharePanel.jsx:156]: Why not have this block here with another `||` operator https://github.com/geosolutions-it/MapStore2/blob/master/web/client/components/share/SharePanel.jsx#L151? We could avoid a separate if block.

**Review by dsuren1 — CHANGES_REQUESTED** [review]
@tonym3nsa 
Could you check this please

**Review by dsuren1 — APPROVED** [review]
LGTM

**Comment by dsuren1** [comment]
@ElenaGallo Kindly test this fix once deployed to DEV

**Comment by ElenaGallo** [comment]
@tonym3nsa there is a regression, now the Min and Max values on Heading, Roll and Pitch allow values beyond the specified limit range
value

**Comment by tonym3nsa** [comment]
@ElenaGallo I can't seem able to recreate the issue; can you please let me know how you entered the figures?

**Comment by ElenaGallo** [comment]
@tonym3nsa you are right, the problem only occurs on the second attempt to enter the value.


3d

**Comment by tdipisa** [comment]
> @ElenaGallo I can't seem able to recreate the issue; can you please let me know how you entered the figures?

The same for me @ElenaGallo

**Comment by tonym3nsa** [comment]
@ElenaGallo I might have seen what the issue is. It happens when you type fast. Providing a  fix right now

File: web/client/components/share/SharePanel.jsx (hunk 1 above)
Comment (tonym3nsa) [reply web/client/components/share/SharePanel.jsx]: @dsuren1 I think I might have to revert to the original implementation. After refactoring to this, the inputted values go beyond the specified range in certain cases.

**Comment by tdipisa** [comment]
@ElenaGallo @tonym3nsa (FYI @dsuren1) It seems to me one problem can be the following. @tonym3nsa (@dsuren1) let me know your thoughts.

Here I'm in a situation (without typing anything in the fields) where values in form seems wrong:

2022_03_08_12_36_31_MapStore_HomePage

and they don't match the ones in the URL (**which anyway shows exactly what I'm trying to share**):

https://dev-mapstore.geosolutionsgroup.com/mapstore/#/viewer/cesium/10358?center=-102.56493882780978,32.27312307744788&zoom=4.636726959224782&heading=6.2471550569244325&pitch=-1.4768349282484725&roll=6.282676334180332
 
image

Below the steps to reproduce (sometimes I can reproduce and sometimes not):

https://demo.geo-solutions.it/share/tobia/3d_cesium.gif

**Comment by dsuren1** [comment]
@tdipisa 
This url https://dev-mapstore.geosolutionsgroup.com/mapstore/#/viewer/cesium/10358?center=-102.56493882780978,32.27312307744788&zoom=4.636726959224782&heading=6.2471550569244325&pitch=-1.4768349282484725&roll=6.282676334180332 has incorrect roll (360), probably created before the fixes, not sure.

Can you try the same from the share panel, by supplying incorrect orientation values, because `onBlur` of the field, the validation kicks in and resets the incorrect value in the field. So it shouldn't be possible to create an incorrect share url with orientations, if not then this needs to be fixed.

**Comment by tdipisa** [comment]
> This url https://dev-mapstore.geosolutionsgroup.com/mapstore/#/viewer/cesium/10358?center=-102.56493882780978,32.27312307744788&zoom=4.636726959224782&heading=6.2471550569244325&pitch=-1.4768349282484725&roll=6.282676334180332 has incorrect roll (360), probably created before the fixes, not sure.
> 
> Can you try the same from the share panel, by supplying incorrect orientation values, because `onBlur` of the field, the validation kicks in and resets the incorrect value in the field. So it shouldn't be possible to create an incorrect share url with orientations, if not then this needs to be fixed.

@dsuren1 as you can see from the video, I've used the URL as it is provided by the Share tool. I disn't change anything in the form, I've simply changed the orientation in Cesium and this sometimes generate wrong values for Roll. In addition, all values in the form are different from the ones in the URL.

**Comment by tonym3nsa** [comment]
@tdipisa the values in the URL are in radian whereas the values in the SharePanel are in degrees - that is what they don't match. What I do is get the value from Cesium and convert it into degrees in the share panel. At times, during the back and forth conversions of the float, instead of getting a whole number such as  for example - 360 degrees, the conversion causes the figure to have more decimal places - thus, in your example, instead of 360 degrees, we have 359.97....

**Comment by tdipisa** [comment]
> @tdipisa the values in the URL are in radian whereas the values in the SharePanel are in degrees - that is why they don't match. What I do is get the radian value from Cesium and convert it into degrees in the share panel. At times, during the back and forth conversions of the float, instead of getting a whole number such as for example - 360 degrees, the conversion causes the figure to have more decimal places - thus, in your example, instead of 360 degrees, we have 359.97....

I understand, thank you for the explanation.

**Pull request merged**